File: app/src/main/java/net/mbonnin/arcanetracker/hsreplay/HSReplay.java
package net.mbonnin.arcanetracker.hsreplay;

import android.os.Build;

import com.google.gson.Gson;

import net.mbonnin.arcanetracker.ArcaneTrackerApplication;
import net.mbonnin.arcanetracker.BuildConfig;
import net.mbonnin.arcanetracker.MainViewCompanion;
import net.mbonnin.arcanetracker.Settings;
import net.mbonnin.arcanetracker.Utils;
import net.mbonnin.arcanetracker.hsreplay.model.Token;
import net.mbonnin.arcanetracker.hsreplay.model.TokenRequest;
import net.mbonnin.arcanetracker.hsreplay.model.UploadRequest;
import net.mbonnin.arcanetracker.model.GameSummary;
import net.mbonnin.arcanetracker.parser.Game;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;

import io.paperdb.Paper;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import rx.Observable;
import rx.Observer;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Created by martin on 11/29/16.
 */

public class HSReplay {
    private static final java.lang.String KEY_GAME_LIST = "KEY_GAME_LIST";
    private static HSReplay sHSReplay;
    private final OkHttpClient mS3Client;
    private final String mUserAgent;
    private String mToken;
    private ArrayList<GameSummary> mGameList;
    private Service mService;

    private final Observer<Token> mTokenObserver = new Observer<Token>() {
        @Override
        public void onCompleted() {

        }

        @Override
        public void onError(Throwable e) {
            Timber.e("createTokenError" + e);
        }

        @Override
        public void onNext(Token token) {
            mToken = token.key;
            Timber.w("got token=" + mToken);
            Settings.set(Settings.HSREPLAY_TOKEN, mToken);
        }
    };
    private Observer<? super Void> mUploadObserver = new Observer<Void>() {
        @Override
        public void onCompleted() {

        }

        @Override
        public void onError(Throwable e) {
            Timber.e(e);
        }

        @Override
        public void onNext(Void aVoid) {
            Timber.w("HSREplay upload success");
        }
    };

    public static HSReplay get() {
        if (sHSReplay == null) {
            sHSReplay = new HSReplay();
        }

        return sHSReplay;
    }

    public ArrayList<GameSummary> getGameSummary() {
        return mGameList;
    }

    public void uploadGameDebug(String matchStart, String friendlyPlayerId, GameSummary summary, String gameStr) {
        UploadRequest uploadRequest = new UploadRequest();
        uploadRequest.match_start = matchStart;
        uploadRequest.build = 20022;
        uploadRequest.friendly_player_id = friendlyPlayerId;
        uploadRequest.game_type = summary.bnetGameType;

        service().createUpload("https://upload.hsreplay.net/api/v1/replay/upload/request", uploadRequest)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .flatMap(upload -> {
                    Timber.w("url is " + upload.url);
                    Timber.w("put_url is " + upload.put_url);

                    if (upload.put_url == null) {
                        return null;
                    }

                    if (summary != null) {
                        summary.hsreplayUrl = upload.url;
                        Paper.book().write(KEY_GAME_LIST, mGameList);
                    }
                    return putToS3(upload.put_url, gameStr).subscribeOn(Schedulers.io());
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(mUploadObserver);
    }

    public void uploadGame(String matchStart, Game game, String gameStr) {
        Timber.w("uploadGame");
        if (game == null) {
            return;
        }

        GameSummary summary = new GameSummary();
        summary.coin = game.getPlayer().hasCoin;
        summary.win = game.victory;
        summary.hero = game.player.classIndex();
        summary.opponentHero = game.opponent.classIndex();
        summary.date = Utils.ISO8601DATEFORMAT.format(new Date());
        summary.deckName = MainViewCompanion.getPlayerCompanion().getDeck().name;
        summary.bnetGameType = game.bnetGameType;

        mGameList.add(0, summary);
        Paper.book().write(KEY_GAME_LIST, mGameList);

        if (mToken == null) {
            return;
        }

        if (Settings.get(Settings.HSREPLAY, Settings.DEFAULT_HSREPLAY)) {
            uploadGameDebug(matchStart, game.player.entity.PlayerID, summary, gameStr);
        }
    }

    public Observable<Void> putToS3(String putUrl, String gameStr) {
        return Observable.fromCallable(() -> {
            RequestBody body = RequestBody.create(null, gameStr);
            Request request = new Request.Builder()
                    .put(body)
                    .url(putUrl)
                    .header("Content-Type", "text/plain")
                    .addHeader("User-Agent", mUserAgent)
                    .build();

            try {
                Response response = mS3Client.newCall(request).execute();
                if (!response.isSuccessful()) {
                    return null;
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        });
    }

    public HSReplay() {
        mGameList = Paper.book().read(KEY_GAME_LIST);
        if (mGameList == null) {
            mGameList = new ArrayList<>();
        }

        mUserAgent = ArcaneTrackerApplication.getContext().getPackageName() + "/" + BuildConfig.VERSION_NAME
                + "; Android " + Build.VERSION.RELEASE + "; " + String.format("%.3f", Utils.getDiagonal()) + "inches";

        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(chain -> {
                    Request request = chain.request();

                    Request.Builder requestBuilder = request.newBuilder();
                    requestBuilder.addHeader("X-Api-Key", "8b27e53b-0256-4ff1-b134-f531009c05a3");
                    requestBuilder.addHeader("User-Agent", mUserAgent);
                    if (mToken != null) {
                        requestBuilder.addHeader("Authorization", "Token " + mToken);
                    }
                    request = requestBuilder.build();

                    return chain.proceed(request);
                }).build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://hsreplay.net/api/v1/")
                .addCallAdapterFactory(RxJavaCallAdapterFactory.createWithScheduler(Schedulers.io()))
                .addConverterFactory(GsonConverterFactory.create(new Gson()))
                .client(client)
                .build();

        mService = retrofit.create(Service.class);

        mS3Client = new OkHttpClient.Builder()
                //.addInterceptor(new GzipRequestInterceptor())
                .build();

        if (Settings.get(Settings.HSREPLAY, Settings.DEFAULT_HSREPLAY)) {
            mToken = Settings.get(Settings.HSREPLAY_TOKEN, null);

            Timber.w("init token=" + mToken);

            if (mToken == null) {
                generateToken();
            }
        }

        //uploadGameDebug(Utils.ISO8601DATEFORMAT.format(new Date()), "1", null, "toto");
    }

    private void generateToken() {
        TokenRequest tokenRequest = new TokenRequest();
        tokenRequest.test_data = Utils.isAppDebuggable() ? true : false;

        service().createToken(tokenRequest)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(mTokenObserver);
    }

    public String getToken() {
        return mToken;
    }

    public void eraseGameSummary() {
        mGameList.clear();
        mToken = null;
        Settings.set(Settings.HSREPLAY_TOKEN, null);
        Paper.book().write(KEY_GAME_LIST, mGameList);

        generateToken();
    }

    public Service service() {
        return mService;
    }
}

File: app/src/main/java/net/mbonnin/arcanetracker/hsreplay/HSReplay.kt
package net.mbonnin.arcanetracker.hsreplay

import android.content.Context
import android.preference.PreferenceManager
import androidx.core.content.edit
import com.google.gson.Gson
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.*
import net.mbonnin.arcanetracker.Utils
import net.mbonnin.arcanetracker.hsreplay.model.Account
import net.mbonnin.arcanetracker.hsreplay.model.legacy.UploadRequest
import net.mbonnin.arcanetracker.hsreplay.model.legacy.UploadToken
import net.mbonnin.arcanetracker.hsreplay.model.legacy.UploadTokenRequest
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import ru.gildor.coroutines.retrofit.await
import timber.log.Timber
import java.io.IOException

class HSReplay(val context: Context, val userAgent: String) {
    private var legacyToken: String? = null

    private val S3Client by lazy {
        OkHttpClient.Builder()
                .addInterceptor(GzipInterceptor())
                .build()
    }
    private val mOauthervice by lazy {
        val oauthOkHttpClient = OkHttpClient.Builder()
                .addInterceptor(HsReplayInterceptor())
                .build()

        Retrofit.Builder()
                .baseUrl("https://api.hsreplay.net/v1/")
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io()))
                .addConverterFactory(GsonConverterFactory.create(Gson()))
                .client(oauthOkHttpClient)
                .build()
                .create(HsReplayService::class.java)
    }
    private val legacyService by lazy {
        val legacyClient = OkHttpClient.Builder()
                .addInterceptor(LegacyInterceptor(userAgent))
                .build()

        Retrofit.Builder()
                .baseUrl("https://hsreplay.net/api/v1/")
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io()))
                .addConverterFactory(GsonConverterFactory.create(Gson()))
                .client(legacyClient)
                .build()
                .create(LegacyService::class.java)
    }
    val sharedPreferences by lazy {
        PreferenceManager.getDefaultSharedPreferences(context)
    }

    init {

        legacyToken = sharedPreferences.getString(KEY_HSREPLAY_LEGACY_TOKEN, null)
        Timber.w("init token=${legacyToken != null}")

        if (HsReplayInterceptor.refreshToken != null) {
            GlobalScope.launch {
                if (legacyToken == null) {
                    // This should never happen as the legacyToken must be set at login time
                    Utils.reportNonFatal(Exception("no token for logged user ?"))
                    createLegacyToken().getOrNull()?.let {
                        legacyToken = it
                        sharedPreferences.edit {
                            putString(KEY_HSREPLAY_LEGACY_TOKEN, it)
                        }
                    }
                }
                val accountResult = account()
                accountResult.getOrNull()?.let {
                    sharedPreferences.edit {
                        putBoolean(KEY_HSREPLAY_PREMIUM, it.is_premium ?: false)
                        putString(KEY_HSREPLAY_BATTLETAG, it.battletag)
                        putString(KEY_HSREPLAY_USERNAME, it.username)
                    }
                }
            }
        }
    }

    suspend fun uploadGame(uploadRequest: UploadRequest, gameStr: String): Result<String> {

        Timber.w("uploadGame [token=$legacyToken]")

        if (legacyToken == null) {
            return Result.failure(Exception("no token"))
        }

        Timber.w("doUploadGame")

        val authorization = "Token $legacyToken"

        val upload = try {
            legacyService.createUpload("https://upload.hsreplay.net/api/v1/replay/upload/request", uploadRequest, authorization).await()
        } catch (e: Exception) {
            return Result.failure(e)
        }

        Timber.w("url is ${upload.url}")
        Timber.w("put_url is ${upload.put_url}")

        val exception = withContext(Dispatchers.IO) {
            putToS3(upload.put_url, gameStr).exceptionOrNull()
        }
        if (exception != null) {
            return Result.failure(exception)
        }

        return Result.success(upload.url!!)
    }

    private fun putToS3(putUrl: String?, gameStr: String): Result<Unit> {

        if (putUrl == null) {
            return Result.failure(Exception("not put url"))
        }

        val body = RequestBody.create(MediaType.parse("text/plain"), gameStr)
        val request = Request.Builder()
                .put(body)
                .url(putUrl)
                .addHeader("User-Agent", userAgent)
                .build()

        try {
            val response = S3Client.newCall(request).execute()
            if (!response.isSuccessful) {
                return Result.failure(Exception("response not successful"))
            }
        } catch (e: IOException) {
            e.printStackTrace()
            return Result.failure(e)
        }

        return Result.success(Unit)
    }

    suspend fun uploadToken(): Result<UploadToken> {
        try {
            return Result.success(legacyService.getToken(legacyToken!!).await())
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    suspend fun account(): Result<Account> {
        return try {
            Result.success(mOauthervice.account().await())
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    private suspend fun claimToken(legacyToken: String): Result<Unit> = coroutineScope {
        val str = "{\"token\": \"$legacyToken\"}"
        val claim = async(Dispatchers.IO) {
            try {
                // The coroutines extensions for retrofit will fail on an empty body
                // https://github.com/JakeWharton/retrofit2-kotlin-coroutines-adapter/issues/5
                val response = mOauthervice.claimToken(RequestBody.create(MediaType.parse("application/json"), str)).execute()
                if (!response.isSuccessful) {
                    throw Exception("Cannot claim token: ${response.code()}")
                }
                Result.success(Unit)
            } catch (e: Exception) {
                Result.failure<Unit>(e)
            }
        }

        claim.await()
    }

    suspend fun createLegacyToken(): Result<String> {
        val tokenRequest = UploadTokenRequest()

        return try {
            Result.success(legacyService.createToken(tokenRequest).await().key!!)
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    // Not sure how different this is from username
    fun battleTag() = sharedPreferences.getString(KEY_HSREPLAY_BATTLETAG, null)

    fun username() = sharedPreferences.getString(KEY_HSREPLAY_USERNAME, null)
    fun isPremium() = sharedPreferences.getBoolean(KEY_HSREPLAY_PREMIUM, false)

    fun token(): String? {
        return legacyToken
    }

    fun logout() {
        HsReplayInterceptor.logout()
        legacyToken = null
        sharedPreferences.edit {
            remove(KEY_HSREPLAY_LEGACY_TOKEN)
            remove(KEY_HSREPLAY_PREMIUM)
            remove(KEY_HSREPLAY_BATTLETAG)
            remove(KEY_HSREPLAY_USERNAME)
        }
    }


    suspend fun login(code: String): Result<Unit> = coroutineScope {
        val tokenDeferred = async {
            createLegacyToken()
        }

        val oauthResult = HsReplayInterceptor.login(code)
        if (oauthResult.isFailure) {
            return@coroutineScope oauthResult.map { Unit }
        }
        val tokenResult = tokenDeferred.await()
        if (tokenResult.isFailure) {
            HsReplayInterceptor.logout()
            return@coroutineScope tokenResult.map { Unit }
        }

        val claimResult = claimToken(tokenResult.getOrNull()!!)
        if (claimResult.isFailure) {
            HsReplayInterceptor.logout()
            return@coroutineScope claimResult.map { Unit }
        }

        val accountResult = account()
        if (accountResult.isFailure) {
            HsReplayInterceptor.logout()
            return@coroutineScope accountResult.map { Unit }
        }
        accountResult.getOrNull()!!.let {
            sharedPreferences.edit {
                legacyToken = tokenResult.getOrNull()!!

                putString(KEY_HSREPLAY_LEGACY_TOKEN, legacyToken!!)

                putBoolean(KEY_HSREPLAY_PREMIUM, it.is_premium ?: false)
                putString(KEY_HSREPLAY_BATTLETAG, it.battletag)
                putString(KEY_HSREPLAY_USERNAME, it.username)
            }
        }
        return@coroutineScope Result.success(Unit)
    }


    companion object {
        const val KEY_HSREPLAY_LEGACY_TOKEN = "HSREPLAY_TOKEN"
        const val KEY_HSREPLAY_PREMIUM = "HSREPLAY_PREMIUM"
        const val KEY_HSREPLAY_BATTLETAG = "HSREPLAY_BATTLETAG"
        const val KEY_HSREPLAY_USERNAME = "HSREPLAY_USERNAME"
    }
}
